pkg/volume/rook/rook.go

```diff
@@ -0,0 +1,334 @@
+
+	err = b.mounter.Mount(globalPDPath, dir, "", options)
+	if err != nil {
+		notMnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+			return err
+		}
+		if !notMnt {
+			if mntErr = b.mounter.Unmount(dir); mntErr != nil {
+				glog.Errorf("Failed to unmount: %v", mntErr)
+				return err
+			}
+			notMnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+			if mntErr != nil {
+				glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+				return err
+			}
+			if !notMnt {
+				// This is very odd, we don't expect it.  We'll try again next sync loop.
+				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", dir)
+				return err
+			}
+		}
+		os.Remove(dir)
+		glog.Errorf("Mount of disk %s failed: %v", dir, err)
+		return err
+	}
+
+	if !b.readOnly {
+		volume.SetVolumeOwnership(b, fsGroup)
+	}
+
+	glog.V(4).Infof("Successfully mounted %s", dir)
+	return nil
+}
+
+type rookUnmounter struct {
+	*rookVolume
+}
+
+var _ volume.Unmounter = &rookUnmounter{}
+
+func (plugin *rookPlugin) NewUnmounter(volName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(volName, podUID)
+}
+
+func (plugin *rookPlugin) newUnmounterInternal(volName string, podUID types.UID) (volume.Unmounter, error) {
+	return &rookUnmounter{&rookVolume{
+		podUID:          podUID,
+		pvName:          volName,
+		plugin:          plugin,
+		mounter:         plugin.host.GetMounter(),
+		MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, volName, plugin.host)),
+	}}, nil
+}
+
+func (c *rookUnmounter) GetPath() string {
+	return getPath(c.podUID, c.pvName, c.plugin.host)
+}
+
+func (c *rookUnmounter) TearDown() error {
+	return c.TearDownAt(c.GetPath())
+}
+
+// TearDownAt unmounts the bind mount
+func (c *rookUnmounter) TearDownAt(dir string) error {
+	return util.UnmountPath(dir, c.mounter)
+}
+
+func (plugin *rookPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	mounter := plugin.host.GetMounter()
+	pluginDir := plugin.host.GetPluginDir(plugin.GetPluginName())
+	sourceName, err := mounter.GetDeviceNameFromMount(mountPath, pluginDir)
+	if err != nil {
+		return nil, err
+	}
+	// sourceName is in the form of xxx/storageclass-group-id
+	devName := path.Base(sourceName)
+	volumeSource := rstring.Split(devName, "-")
+
+	rookVolume := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				Rook: &v1.RookVolumeSource{
+					VolumeGroup: volumeSource[1],
+					VolumeID:    volumeSource[2],
+				},
+			},
+			StorageClassName: volumeSource[0],
+		},
+	}
+	return volume.NewSpecFromPersistentVolume(rookVolume, false), nil
+}
+
+func (plugin *rookPlugin) RequiresRemount() bool {
+	return false
+}
+
+func (plugin *rookPlugin) SupportsMountOption() bool {
```

Comment:
btw to support mount options for rook, you will have to also add Rook volume here - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/validation/pv_validation.go#L50